static/app/views/profiling/profileDetails.tsx

```diff
@@ -149,6 +148,127 @@ function ProfileDetails() {
     [location, search]
   );
 
+  const [filters, setFilters] = useState<Partial<Record<TableColumnKey, string[]>>>({});
+
+  const columnFilters = useMemo(() => {
+    function makeOnFilterChange(key: string) {
+      return values => {
+        setFilters(prevFilters => ({
+          ...prevFilters,
+          [key]: values.length > 0 ? values.map(val => val.value) : undefined,
+        }));
+      };
+    }
+    return {
+      type: {
+        values: ['application', 'system'],
+        onChange: makeOnFilterChange('type'),
+      },
+      image: {
+        values: pluckUniqueValues(slowestFunctions, 'image').sort((a, b) =>
+          a.localeCompare(b)
+        ),
+        onChange: makeOnFilterChange('image'),
+      },
+    };
+  }, [slowestFunctions]);
+
+  const currentSort = useMemo<GridColumnSortBy<TableColumnKey>>(() => {
+    let key = location.query?.functionsSort ?? '';
+
+    const defaultSort = {
+      key: 'self weight',
+      order: 'desc',
+    } as GridColumnSortBy<TableColumnKey>;
+
+    const isDesc = key[0] === '-';
+    if (isDesc) {
+      key = key.slice(1);
+    }
+
+    if (!key || !tableColumnKey.includes(key as TableColumnKey)) {
+      return defaultSort;
+    }
+
+    return {
+      key,
+      order: isDesc ? 'desc' : 'asc',
+    } as GridColumnSortBy<TableColumnKey>;
+  }, [location.query]);
+
+  useEffect(() => {
+    const removeListener = browserHistory.listenBefore((nextLocation, next) => {
+      if (location.pathname === nextLocation.pathname) {
+        next(nextLocation);
+        return;
+      }
+
+      if ('functionsSort' in nextLocation.query) {
+        delete nextLocation.query.functionsSort;
+      }
+
+      next(nextLocation);
+    });
+
+    return removeListener;
+  });
+
+  const generateSortLink = useCallback(
+    (column: TableColumnKey) => {
+      if (!SORTABLE_COLUMNS.has(column)) {
+        return () => undefined;
+      }
+      if (!currentSort) {
+        return () => ({
+          ...location,
+          query: {
+            ...location.query,
+            functionsSort: column,
+          },
+        });
+      }
+
+      const direction =
+        currentSort.key !== column
+          ? 'desc'
+          : currentSort.order === 'desc'
+          ? 'asc'
+          : 'desc';
+
+      return () => ({
+        ...location,
+        query: {
+          ...location.query,
+          functionsSort: `${direction === 'desc' ? '-' : ''}${column}`,
+        },
+      });
+    },
+    [location, currentSort]
+  );
+
+  const data = slowestFunctions
+    .filter(row => {
+      let include = true;
+      for (const key in filters) {
+        const values = filters[key];
+        if (!values) {
+          continue;
+        }
+        include = values.includes(row[key]);
+        if (!include) {
+          return false;
```

Comment:
This will exit the loop on the first filter if it does not include the value, do you want to `continue` here instead to align with L255. If not, I would add a comment here why we are exiting on the first filter as it's not immediately obvious.

Reply:
sorry, just need to make it more clear. L254:255 are implicitly doing an undefined check, if no filter is set we just wanna continue.

the row must match all filters in order for it to be included in the table; hence we return false.

fwiw; doing this crap in a loop is overkill rn for what it is since we're only checking two filters. all this stuff needs work to be made more of a common general util. similar to the issues faced w/ qs, i just don't want to get too off track on the PR and would rather refine it after.

there's alot of value in improving our grid/table utils... everything is pretty manual and imperative right now.